src/libopenrave-core/environment-core.h

```diff
@@ -2986,181 +2986,170 @@ class Environment : public EnvironmentBase
+                pExistingBody = std::move(vBodies[existingBodyIndex]); // Invalidate this slot in vBodies to make sure we don't accidentally re-use it
+                OPENRAVE_ASSERT_OP(pExistingBody, !=, KinBodyPtr()); // Only valid entries should exist in our lookup maps
+                existingBodyIndicesById.erase(pExistingBody->GetId()); // Also ensures no dangling string_views in the event this body gets destroyed later in the loop
+                existingBodyIndicesByName.erase(pExistingBody->GetName());
+
+                // If we matched an existing body, but that body is of a different interface to the info (e.g robot vs plain body), then we can't just update, need to recreate it.
+                bool bInterfaceMatches = pExistingBody->GetXMLId() == pKinBodyInfo->_interfaceType;
+                if (!bInterfaceMatches || pExistingBody->IsRobot() != pKinBodyInfo->_isRobot) {
+                    KinBodyPtr pBodyToRemove = std::move(pExistingBody); // Invalidates existing body output
+                    RAVELOG_VERBOSE_FORMAT("env=%s, body '%s' interface is changed, remove old body from environment. xmlid=%s, _interfaceType=%s, isRobot %d != %d", GetNameId() % pBodyToRemove->_id % pBodyToRemove->GetXMLId() % pKinBodyInfo->_interfaceType % pBodyToRemove->IsRobot() % pKinBodyInfo->_isRobot);
+                    vRemovedBodies.push_back(pBodyToRemove);
+
+                    ExclusiveLock lock690(_mutexInterfaces);
+                    vector<KinBodyPtr>::iterator itBodyToRemove = std::find(_vecbodies.begin(), _vecbodies.end(), pBodyToRemove);
+                    if (itBodyToRemove != _vecbodies.end()) {
+                        _InvalidateKinBodyFromEnvBodyIndex(pBodyToRemove->GetEnvironmentBodyIndex());
                     }
+                }
 
-                    if (updateMode == UFIM_OnlySpecifiedBodiesExact) {
-                        usedBodyIndexSet.emplace(nMatchingIndex);
-                    }
+                // If we matched by ID instead of name, it's possible there exists another body in the env with the same name as our body info.
+                // This would cause a conflict if we try and update our current (id matched body) to have the same name.
+                // Since the other body with the same name might get processed again later (by id?), temporarily rename it so that we can continue.
+                if (!!pExistingBody && existingBodyIndex != existingBodyIndexSameName && existingBodyIndexSameName >= 0) {
+                    const KinBodyPtr& pExistingBodySameName = vBodies[existingBodyIndexSameName];
+                    RAVELOG_DEBUG_FORMAT("env=%s, have to clear body name '%s' id=%s for loading body with id=%s", GetNameId() % pExistingBodySameName->GetName() % pExistingBodySameName->GetId() % pExistingBody->GetId());
+                    pExistingBodySameName->SetName(_GetUniqueName(pExistingBodySameName->GetName() + "_tempRenamedDueToConflict_"));
```

Comment:
@rschlaikjer  Changing name of an existing body invalidates the `string_view` you are managing here, which could corrupt your state.

```c++
        std::unordered_set<string_view> bodyIdsToUpdate; // string_views are over info._vBodyInfos, guaranteed lifetime for the rest of this function
        std::unordered_set<string_view> bodyNamesToUpdate;
```

Reply:
There are also these variables that use string_view which might also be affected:
```c++
                const std::unordered_map<string_view, int>::iterator itExistingBodyIndexById = existingBodyIndicesById.find(kinBodyInfo._id);
                const std::unordered_map<string_view, int>::iterator itExistingBodyIndexByName = existingBodyIndicesByName.find(kinBodyInfo._name);
```

Reply:
Those first views are over the body _info_ record name strings (`info._vBodyInfos`, per the comment), so they don't need to worry about the data getting changed (info is immutable for this function).

As for the second one, good catch - we handle removing the reused body, but not any fallout from renaming bodies.